{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "execution": {
     "iopub.execute_input": "2023-05-14T03:23:42.546940Z",
     "iopub.status.busy": "2023-05-14T03:23:42.546505Z",
     "iopub.status.idle": "2023-05-14T03:23:57.425383Z",
     "shell.execute_reply": "2023-05-14T03:23:57.423299Z"
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: Flask in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (2.2.3)\r\n",
      "Requirement already satisfied: importlib-metadata>=3.6.0 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from Flask) (4.11.3)\r\n",
      "Requirement already satisfied: itsdangerous>=2.0 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from Flask) (2.0.1)\r\n",
      "Requirement already satisfied: Werkzeug>=2.2.2 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from Flask) (2.2.3)\r\n",
      "Requirement already satisfied: Jinja2>=3.0 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from Flask) (3.1.2)\r\n",
      "Requirement already satisfied: click>=8.0 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from Flask) (8.0.4)\r\n",
      "Requirement already satisfied: zipp>=0.5 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from importlib-metadata>=3.6.0->Flask) (3.8.0)\r\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from Jinja2>=3.0->Flask) (2.1.2)\r\n",
      "\r\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.0.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.1.2\u001b[0m\r\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\r\n",
      "Requirement already satisfied: requests in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (2.28.1)\r\n",
      "Requirement already satisfied: charset-normalizer<3,>=2 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from requests) (2.0.4)\r\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from requests) (1.26.11)\r\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from requests) (2022.9.24)\r\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from requests) (3.3)\r\n",
      "\r\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.0.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.1.2\u001b[0m\r\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\r\n",
      "Requirement already satisfied: google-cloud-bigquery in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (3.10.0)\r\n",
      "Requirement already satisfied: google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0dev,>=1.31.5 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-cloud-bigquery) (2.11.0)\r\n",
      "Requirement already satisfied: python-dateutil<3.0dev,>=2.7.2 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-cloud-bigquery) (2.8.2)\r\n",
      "Requirement already satisfied: requests<3.0.0dev,>=2.21.0 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-cloud-bigquery) (2.28.1)\r\n",
      "Requirement already satisfied: proto-plus<2.0.0dev,>=1.15.0 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-cloud-bigquery) (1.22.2)\r\n",
      "Requirement already satisfied: packaging>=20.0.0 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-cloud-bigquery) (21.3)\r\n",
      "Requirement already satisfied: google-cloud-core<3.0.0dev,>=1.6.0 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-cloud-bigquery) (2.3.2)\r\n",
      "Requirement already satisfied: grpcio<2.0dev,>=1.47.0 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-cloud-bigquery) (1.54.2)\r\n",
      "Requirement already satisfied: google-resumable-media<3.0dev,>=0.6.0 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-cloud-bigquery) (2.5.0)\r\n",
      "Requirement already satisfied: protobuf!=3.20.0,!=3.20.1,!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.19.5 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-cloud-bigquery) (4.23.0)\r\n",
      "Requirement already satisfied: googleapis-common-protos<2.0dev,>=1.56.2 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0dev,>=1.31.5->google-cloud-bigquery) (1.59.0)\r\n",
      "Requirement already satisfied: google-auth<3.0dev,>=2.14.1 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0dev,>=1.31.5->google-cloud-bigquery) (2.16.0)\r\n",
      "Requirement already satisfied: grpcio-status<2.0dev,>=1.33.2 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0dev,>=1.31.5->google-cloud-bigquery) (1.54.2)\r\n",
      "Requirement already satisfied: google-crc32c<2.0dev,>=1.0 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-resumable-media<3.0dev,>=0.6.0->google-cloud-bigquery) (1.5.0)\r\n",
      "Requirement already satisfied: pyparsing!=3.0.5,>=2.0.2 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from packaging>=20.0.0->google-cloud-bigquery) (3.0.9)\r\n",
      "Requirement already satisfied: six>=1.5 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from python-dateutil<3.0dev,>=2.7.2->google-cloud-bigquery) (1.16.0)\r\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from requests<3.0.0dev,>=2.21.0->google-cloud-bigquery) (3.3)\r\n",
      "Requirement already satisfied: charset-normalizer<3,>=2 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from requests<3.0.0dev,>=2.21.0->google-cloud-bigquery) (2.0.4)\r\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from requests<3.0.0dev,>=2.21.0->google-cloud-bigquery) (1.26.11)\r\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from requests<3.0.0dev,>=2.21.0->google-cloud-bigquery) (2022.9.24)\r\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-auth<3.0dev,>=2.14.1->google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0dev,>=1.31.5->google-cloud-bigquery) (0.2.8)\r\n",
      "Requirement already satisfied: rsa<5,>=3.1.4 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-auth<3.0dev,>=2.14.1->google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0dev,>=1.31.5->google-cloud-bigquery) (4.9)\r\n",
      "Requirement already satisfied: cachetools<6.0,>=2.0.0 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-auth<3.0dev,>=2.14.1->google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0dev,>=1.31.5->google-cloud-bigquery) (5.3.0)\r\n",
      "Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from pyasn1-modules>=0.2.1->google-auth<3.0dev,>=2.14.1->google-api-core[grpc]!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0dev,>=1.31.5->google-cloud-bigquery) (0.4.8)\r\n",
      "\r\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.0.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.1.2\u001b[0m\r\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\r\n",
      "Requirement already satisfied: google-auth in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (2.16.0)\r\n",
      "Requirement already satisfied: google-auth-oauthlib in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (0.4.6)\r\n",
      "Requirement already satisfied: google-auth-httplib2 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (0.1.0)\r\n",
      "Requirement already satisfied: google-api-python-client in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (2.86.0)\r\n",
      "Requirement already satisfied: cachetools<6.0,>=2.0.0 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-auth) (5.3.0)\r\n",
      "Requirement already satisfied: pyasn1-modules>=0.2.1 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-auth) (0.2.8)\r\n",
      "Requirement already satisfied: rsa<5,>=3.1.4 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-auth) (4.9)\r\n",
      "Requirement already satisfied: six>=1.9.0 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-auth) (1.16.0)\r\n",
      "Requirement already satisfied: requests-oauthlib>=0.7.0 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-auth-oauthlib) (1.3.1)\r\n",
      "Requirement already satisfied: httplib2>=0.15.0 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-auth-httplib2) (0.22.0)\r\n",
      "Requirement already satisfied: uritemplate<5,>=3.0.1 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-api-python-client) (4.1.1)\r\n",
      "Requirement already satisfied: google-api-core!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0dev,>=1.31.5 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-api-python-client) (2.11.0)\r\n",
      "Requirement already satisfied: requests<3.0.0dev,>=2.18.0 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-api-core!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0dev,>=1.31.5->google-api-python-client) (2.28.1)\r\n",
      "Requirement already satisfied: protobuf!=3.20.0,!=3.20.1,!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<5.0.0dev,>=3.19.5 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-api-core!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0dev,>=1.31.5->google-api-python-client) (4.23.0)\r\n",
      "Requirement already satisfied: googleapis-common-protos<2.0dev,>=1.56.2 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from google-api-core!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0dev,>=1.31.5->google-api-python-client) (1.59.0)\r\n",
      "Requirement already satisfied: pyparsing!=3.0.0,!=3.0.1,!=3.0.2,!=3.0.3,<4,>=2.4.2 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from httplib2>=0.15.0->google-auth-httplib2) (3.0.9)\r\n",
      "Requirement already satisfied: pyasn1<0.5.0,>=0.4.6 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from pyasn1-modules>=0.2.1->google-auth) (0.4.8)\r\n",
      "Requirement already satisfied: oauthlib>=3.0.0 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from requests-oauthlib>=0.7.0->google-auth-oauthlib) (3.2.2)\r\n",
      "Requirement already satisfied: charset-normalizer<3,>=2 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from requests<3.0.0dev,>=2.18.0->google-api-core!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0dev,>=1.31.5->google-api-python-client) (2.0.4)\r\n",
      "Requirement already satisfied: urllib3<1.27,>=1.21.1 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from requests<3.0.0dev,>=2.18.0->google-api-core!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0dev,>=1.31.5->google-api-python-client) (1.26.11)\r\n",
      "Requirement already satisfied: idna<4,>=2.5 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from requests<3.0.0dev,>=2.18.0->google-api-core!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0dev,>=1.31.5->google-api-python-client) (3.3)\r\n",
      "Requirement already satisfied: certifi>=2017.4.17 in /Users/chandanasy/opt/anaconda3/lib/python3.9/site-packages (from requests<3.0.0dev,>=2.18.0->google-api-core!=2.0.*,!=2.1.*,!=2.2.*,!=2.3.0,<3.0.0dev,>=1.31.5->google-api-python-client) (2022.9.24)\r\n",
      "\r\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m A new release of pip is available: \u001b[0m\u001b[31;49m23.0.1\u001b[0m\u001b[39;49m -> \u001b[0m\u001b[32;49m23.1.2\u001b[0m\r\n",
      "\u001b[1m[\u001b[0m\u001b[34;49mnotice\u001b[0m\u001b[1;39;49m]\u001b[0m\u001b[39;49m To update, run: \u001b[0m\u001b[32;49mpip install --upgrade pip\u001b[0m\r\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import json\n",
    "os.environ[\"GOOGLE_APPLICATION_CREDENTIALS\"] = \"replace with your credentials\"\n",
    "!pip install Flask\n",
    "!pip install requests\n",
    "!pip install google-cloud-bigquery\n",
    "!pip install google-auth google-auth-oauthlib google-auth-httplib2 google-api-python-client\n",
    "\n",
    "from google.cloud import bigquery\n",
    "\n",
    "# Create a BigQuery client\n",
    "client = bigquery.Client()\n",
    "\n",
    "from collections import defaultdict\n",
    "import datetime\n",
    "import json\n",
    "\n",
    "# Define the date ranges for each president's term in Unix timestamp format\n",
    "obama_start_date = int(datetime.datetime(2009, 1, 20, 0, 0, 0).timestamp())\n",
    "obama_end_date = int(datetime.datetime(2017, 1, 20, 0, 0, 0).timestamp())\n",
    "trump_start_date = int(datetime.datetime(2017, 1, 20, 0, 0, 0).timestamp())\n",
    "trump_end_date = int(datetime.datetime(2021, 1, 20, 0, 0, 0).timestamp())\n",
    "biden_start_date = int(datetime.datetime(2021, 1, 20, 0, 0, 0).timestamp())\n",
    "biden_end_date = int(datetime.datetime.utcnow().timestamp())\n",
    "\n",
    "# Query to filter rows by date ranges and calculate sentiment counts\n",
    "query = f\"\"\"\n",
    "SELECT\n",
    "    CASE\n",
    "        WHEN created_utc >= {obama_start_date} AND created_utc <= {obama_end_date} THEN 'Barack Obama'\n",
    "        WHEN created_utc >= {trump_start_date} AND created_utc <= {trump_end_date} THEN 'Donald Trump'\n",
    "        WHEN created_utc >= {biden_start_date} THEN 'Joe Biden'\n",
    "    END AS president,\n",
    "    CASE\n",
    "        WHEN prediction = 0.0 THEN 'negative'\n",
    "        WHEN prediction = 1.0 THEN 'positive'\n",
    "        WHEN prediction = 2.0 THEN 'neutral'\n",
    "    END AS sentiment,\n",
    "    COUNT(*) AS count\n",
    "FROM\n",
    "    famous-athlete-386604.reddit.comment\n",
    "WHERE\n",
    "    created_utc >= {obama_start_date} AND created_utc <= UNIX_SECONDS(CURRENT_TIMESTAMP())  -- Adjust the date range accordingly\n",
    "GROUP BY\n",
    "    president, sentiment\n",
    "ORDER BY\n",
    "    president, sentiment\n",
    "\"\"\"\n",
    "\n",
    "# Run the query\n",
    "query_job = client.query(query)\n",
    "\n",
    "# Get the results\n",
    "results = query_job.result()\n",
    "\n",
    "# Store the results in a dictionary\n",
    "data = defaultdict(dict)\n",
    "for row in results:\n",
    "    president = row['president']\n",
    "    sentiment = row['sentiment']\n",
    "    count = row['count']\n",
    "    data[president][sentiment] = count\n",
    "\n",
    "# Convert the data dictionary to the desired format\n",
    "sentiment_counts = {\n",
    "    \"Barack Obama\": {\n",
    "        \"negative\": data[\"Barack Obama\"].get(\"negative\", 0),\n",
    "        \"neutral\": data[\"Barack Obama\"].get(\"neutral\", 0),\n",
    "        \"positive\": data[\"Barack Obama\"].get(\"positive\", 0)\n",
    "    },\n",
    "    \"Donald Trump\": {\n",
    "        \"negative\": data[\"Donald Trump\"].get(\"negative\", 0),\n",
    "        \"neutral\": data[\"Donald Trump\"].get(\"neutral\", 0),\n",
    "        \"positive\": data[\"Donald Trump\"].get(\"positive\", 0)\n",
    "    },\n",
    "    \"Joe Biden\": {\n",
    "        \"negative\": data[\"Joe Biden\"].get(\"negative\", 0),\n",
    "        \"neutral\": data[\"Joe Biden\"].get(\"neutral\", 0),\n",
    "        \"positive\": data[\"Joe Biden\"].get(\"positive\", 0)\n",
    "    }\n",
    "}\n",
    "\n",
    "# Clear the contents of sentimentCounts.json if it exists\n",
    "if os.path.exists('sentimentCounts.json'):\n",
    "    os.remove('sentimentCounts.json')\n",
    "\n",
    "# Write the sentiment_counts dictionary to sentimentCounts.json\n",
    "with open('sentimentCounts.json', 'w') as file:\n",
    "    json.dump(sentiment_counts, file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
